modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java

```diff
@@ -479,6 +522,64 @@ private void checkTypesInteroperability(SqlValidatorScope scope, SqlNode expr) {
                     throw SqlUtil.newContextException(expr.getParserPosition(), ex);
                 }
             }
+
+            if (castOp) {
+                literalCanFitType(expr, returnType);
+            }
+        }
+    }
+
+    /** Check literal can fit to declared exact numeric type, work only for single literal. */
+    private void literalCanFitType(SqlNode expr, RelDataType toType) {
+        if (INT_TYPES.contains(toType.getSqlTypeName())) {
+            SqlLiteral literal = Objects.requireNonNullElseGet(litExtractor, LiteralExtractor::new).getLiteral(expr);
+
+            if (literal == null || literal.toValue() == null) {
+                return;
+            }
+
+            long max = Commons.getMaxValue(toType);
```

Comment:
I think it is better to use calcite's `SqlTypeName::getLimit` here see `RexUtils::toSaturatedValue`.